modules/auxiliary/scanner/ssh/ssh_enum_git_keys.rb

```diff
@@ -0,0 +1,181 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Auxiliary::Scanner
+  include Msf::Auxiliary::Report
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name'          => 'Test SSH Github Access',
+        'Description'   => %q(
+          This module will attempt to test remote Git access using
+          (.ssh/id_* private keys). This works against GitHub and
+          GitLab by default, but can easily be extended to support
+          more server types.
+        ),
+        'License'       => MSF_LICENSE,
+        'Author'        => ['Wyatt Dahlenburg (@wdahlenb)'],
+        'Platform'      => ['linux'],
+        'SessionTypes'  => ['shell', 'meterpreter'],
+        'References'    => [['URL', 'https://help.github.com/en/articles/testing-your-ssh-connection']]
+      )
+    )
+
+    register_options(
+      [
+        OptPath.new('KEY_FILE', [false, 'Filename of a private key.', nil]),
+        OptPath.new('KEY_DIR', [false, 'Directory of several keys. Filenames will be recursivley found matching id_* (Ex: /home/user/.ssh)', nil]),
+        OptString.new('GITSERVER', [false, 'Optional parameter to specify alternate Git Server (GitHub, GitLab, etc)', 'github.com'])
+      ]
+    )
+    deregister_options(
+      'RHOST', 'RHOSTS', 'PASSWORD', 'PASS_FILE', 'BLANK_PASSWORDS', 'USER_AS_PASS', 'USERPASS_FILE', 'DB_ALL_PASS', 'DB_ALL_CREDS'
+    )
+
+  end
+
+  # OPTPath will revert to pwd when set back to ""
+  def key_dir
+    datastore['KEY_DIR'] != `pwd`.strip ? datastore['KEY_DIR'] : ""
+  end
+
+  def key_file
+    datastore['KEY_FILE'] != `pwd`.strip ? datastore['KEY_FILE'] : ""
+  end
+
+  def check_key_for_passphrase(file)
+    response = `ssh-keygen -y -P "" -f #{file} 2>&1`
+    return response.include? 'incorrect passphrase'
+  end
+
+  def read_keyfile(file)
+    if file.is_a? Array
+      keys = []
+      file.each do |dir_entry|
+        next unless ::File.readable? dir_entry
+
+        keys.concat(read_keyfile(dir_entry))
+      end
+      return keys
+    else
+      keyfile = ::File.open(file, "rb") { |f| f.read(f.stat.size) }
+    end
+    keys = []
+    this_key = []
+    in_key = false
+    keyfile.split("\n").each do |line|
+      in_key = true if (line =~ /^-----BEGIN ([RD]SA|OPENSSH) PRIVATE KEY-----/)
+      this_key << line if in_key
+      if (line =~ /^-----END ([RD]SA|OPENSSH) PRIVATE KEY-----/)
+        in_key = false
+        keys << file unless check_key_for_passphrase(file)
+      end
+    end
+    if keys.empty?
+      print_error "SSH - No valid keys found"
+    end
+    return keys
+  end
+
+  def provide_user(output)
```

Comment:
`provide_user` is not an intuitive method name.

Reply:
Thoughts on parse_user ?

Reply:
Later on, you reject all keys when no `user` is present with `unless user.empty?`. The only place that the `user` is populated is in this method. This implies that this method might do more than extract/parse the username - it is also solely responsible for determining whether authentication was successful.

Reply:
This was switched to parse_user and the unless statement was removed from the table. Entries will only be added if the username is found and matches the regex per supported git server.